Look for per-project configuration file in parent directories

chutzimir opened the pull request:
Inspired be ESLint's search, it looks for configuration files in all parent directories up until it reaches the user's home or root.

closes #571

adrienverge commented on docs/configuration.rst:
```diff
@@ -15,7 +15,8 @@ If ``-c`` is not provided, yamllint will look for a configuration file in the
 
 - a file named ``.yamllint``, ``.yamllint.yaml``, or ``.yamllint.yml`` in the
-  current working directory
+  current working directory, or a parent directory. The search for this file is
+  terminated at the user's home, or the root.
```
To keep the sentence grammatically correct, I propose:
```suggestion
  current working directory, or a parent directory (the search for this file is
  terminated at the user's home or filesystem root)
```

adrienverge commented on tests/test_cli.py:
```diff
@@ -734,3 +734,27 @@ def test_config_file(self):
+            with self.subTest(conf_file):
+                with temp_workspace(workspace):
+                    with RunContext(self) as ctx:
+                        os.chdir('a/b/c/d/e/f')
```
I was surprised that calling `os.chdir()` here doesn't need to be "undone" after each iteration, but apparently it's OK, it's automatically done by `temp_workspace()`: https://github.com/adrienverge/yamllint/blob/019c87d/tests/common.py#L76-L86


```diff
@@ -741,10 +741,13 @@ class CommandLineConfigTestCase(unittest.TestCase):
                 'extends: relaxed\n')
 
         for conf_file in ('.yamllint', '.yamllint.yml', '.yamllint.yaml'):
+            print('starting test in ' + os.getcwd())
             with self.subTest(conf_file):
                 with temp_workspace(workspace):
+                    print('running subtest in ' + os.getcwd())
                     with RunContext(self) as ctx:
                         os.chdir('a/b/c/d/e/f')
+                        print('running context in ' + os.getcwd())
                         cli.run(('-f', 'parsable', '.'))
 
                 self.assertEqual((ctx.returncode, ctx.stdout, ctx.stderr),
```

chutzimir replied:
Indeed, I checked that temp_workdir goes back to the previous working directory, so this should be good as is I hope.

adrienverge commented on yamllint/cli.py:
```diff
@@ -140,6 +140,18 @@ def show_problems(problems, file, args_format, no_warn):
     return max_level
 
+def find_project_config(path='.'):
+
+    for filename in [ '.yamllint', '.yamllint.yaml', '.yamllint.yml' ]:
```
You can remove the empty line, and also use parentheses rather than brackets, like in the test file.

adrienverge commented on yamllint/cli.py:
```diff
@@ -140,6 +140,18 @@ def show_problems(problems, file, args_format, no_warn):
+        if os.path.isfile(file):
+            return file
+
+    if os.path.abspath(path) == os.path.abspath(os.path.expanduser('~')):
+        return False
+    if os.path.abspath(path) == os.path.abspath('/'):
+        return False
```
- I suggest returning `None` rather than `False`, this is more Pythonic.
- Will the comparison to `os.path.abspath('/')` work on Windows-like systems?

chutzimir replied:
> Will the comparison to `os.path.abspath('/')` work on Windows-like systems?

It seems that it will work...

```
Python 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.getcwd()
'C:\\Program Files\\Python310'
>>> os.path.abspath(os.path.join(os.getcwd(), '..', '..'))
'C:\\'
>>> os.path.abspath('/')
'C:\\'
>>> os.path.abspath(os.path.join(os.getcwd(), '..', '..')) == os.path.abspath('/')
True
```

chutzimir replied:
I addressed all your comments and tried running the tests on Windows, but unfortunately there are too many errors on Windows.

First try:

```
python -m unittest tests/test_cli.py
E
======================================================================
ERROR: test_cli (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_cli
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "C:\Users\georgi.georgiev\work\git\yamllint\tests\test_cli.py", line 17, in <module>
    import fcntl
ModuleNotFoundError: No module named 'fcntl'


----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
```

After deleting the impossible tests, I am getting many errors mostly due to new line format and path separators...

Small sample:
```
First differing element 0:
1
0

- (1,
+ (0,
-  '.\\a.yml:1:1: [warning] missing document start "---" (document-start)\n'
?    ^^

+  './a.yml:1:1: [warning] missing document start "---" (document-start)\n',
?    ^                                                                     +

-  '.\\a.yml:1:13: [error] wrong new line character: expected \\n (new-lines)\n',
   '')
```

```
First differing element 5:
'C:\\[60 chars]3teckb\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\s\\file.yaml'
'C:\\[60 chars]3teckb\\s/s/s/s/s/s/s/s/s/s/s/s/s/s/s/file.yaml'
```

But I think the comparison is good, I even ran yamllint on windows and it doesn't trigger a recursion limit.

adrienverge commented on yamllint/cli.py:
```diff
@@ -192,12 +204,8 @@ def run(argv=None):
-            conf = YamlLintConfig(file='.yamllint.yaml')
-        elif os.path.isfile('.yamllint.yml'):
-            conf = YamlLintConfig(file='.yamllint.yml')
+        elif find_project_config():
+            conf = YamlLintConfig(file=find_project_config())
```
Ideally it would be best to avoid calling `find_project_config()` twice (but also to avoid calling it before the `if/elif` block if not needed). Maybe with a walrus operator `:=`, added in Python 3.8? (but yamllint still supports 3.7...) Or another way?

chutzimir replied:
Let me call the function before going through all the loops, and only load the file in that spot. I noticed that the CI runs against Python 3.7 as well, so did not go for the `:=` which would have been convenient.

adrienverge reviewed (commented):
Hello, thanks for this idea and implementation!

Although I'm favorable to the idea, this is a **breaking change** (behavior would start changing if top-level `.yamllint` files would be taken into account). **Is anyone following this repo opposed to this?**

Can you also add a `test_multiple_parent_config_file()` that checks that if multiple config files have conflicting rules options, the closest one takes precedence?

chutzimir commented:
I addressed all of the above comments, so let me mark them as resolved. I just now noticed the CONTRIBUTING.md file, so I also fixed the linting problems per flake8 advice. I squashed the commits too, to keep the git history clean.

As to whether this is a breaking change or not... I understand your sentiment. I just want to note that I found it very counterintuitive to get different behavior when I edit a file depending on which directory I start from.

By the way, this makes me realize that what I really intended was to find configuration starting from the lint destination, and not starting from the current directory. Let me think about it a bit...

adrienverge commented on yamllint/cli.py:
```diff
@@ -141,6 +141,19 @@ def show_problems(problems, file, args_format, no_warn):
     return max_level
 
 
+def find_project_config_filename(path='.'):
```
I see you rename `find_project_config()` to `find_project_config_filename()`, but since it returns a path I would prefer `find_project_config_path()`, what do you think?

chutzimir replied:
Valid point. I agree. Corrected, squashed, and pushed.

adrienverge commented on yamllint/cli.py:
```diff
@@ -141,6 +141,19 @@ def show_problems(problems, file, args_format, no_warn):
+
+    if os.path.abspath(path) == os.path.abspath(os.path.expanduser('~')):
+        return None
+    if os.path.abspath(path) == os.path.abspath('/'):
+        return None
```
> > Will the comparison to `os.path.abspath('/')` work on Windows-like systems?
> 
> It seems that it will work...
> 
> ```
> Python 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import os
> >>> os.getcwd()
> 'C:\\Program Files\\Python310'
> >>> os.path.abspath(os.path.join(os.getcwd(), '..', '..'))
> 'C:\\'
> >>> os.path.abspath('/')
> 'C:\\'
> >>> os.path.abspath(os.path.join(os.getcwd(), '..', '..')) == os.path.abspath('/')
> True
> ```

OK so it works for `C:\\`. But what if the project is under `D:\\`? I fear a possible infinite loop in some situations.

Maybe a safer implementation would be to check whether `os.path.abspath(path) == os.path.abspath(os.path.join(path, '..'))`, and stop if it's the case?

chutzimir replied:
You're spot on. Corrected, squashed, and pushed.

adrienverge reviewed (commented):
Thanks for the updated version @chutzimir, it looks good. Please see my 2 comments.

Since no one seems to be opposed to this feature (which is a **breaking change**), I'm +1 to merge it.

adrienverge reviewed (approved):
Great! Thanks for contributing to yamllint Georgi :+1:
